src/common/types.ts

```diff
@@ -66,39 +66,66 @@ export type RouteResourceLoading = boolean;
 
 export type RouteResourceTimestamp = number | null;
 
-export type RouteResourceError = Record<string, any> | Error | null;
+export type RouteResourceError = Record<string, any> | Error;
 
-export type RouteResourceData = Record<string, any> | null;
+export type RouteResourceDataPayload = Record<string, any>;
 
-export type RouteResourcePromise = Promise<any> | null;
+export type RouteResourcePromise<T> = Promise<T> | null;
 
-export type RouteResourceUpdater = (
+export type RouteResourceUpdater<RouteResourceData = unknown> = (
   data: RouteResourceData
 ) => RouteResourceData;
 
-export type RouteResourceResponse = {
-  loading: RouteResourceLoading;
+export type RouteResourceResponseBase<RouteResourceData> = {
```

Comment:
this is a very undefined "base" shape, which might be more precisely defined later using `discriminated unions`